src/ng/compile.js

```diff
@@ -2245,43 +2336,62 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
               directiveNormalize(nodeName), 'E', maxPriority, ignoreDirective);
 
           // iterate over the attributes
-          for (var attr, name, nName, ngAttrName, value, isNgAttr, nAttrs = node.attributes,
+          for (var attr, name, nName, value, ngPrefixMatch, nAttrs = node.attributes,
                    j = 0, jj = nAttrs && nAttrs.length; j < jj; j++) {
             var attrStartName = false;
             var attrEndName = false;
 
+            var isNgAttr = false, isNgProp = false, isNgEvent = false;
+            var multiElementMatch;
+
             attr = nAttrs[j];
             name = attr.name;
             value = attr.value;
 
-            // support ngAttr attribute binding
-            ngAttrName = directiveNormalize(name);
-            isNgAttr = NG_ATTR_BINDING.test(ngAttrName);
-            if (isNgAttr) {
+            // support ng-attr-*, ng-prop-* and ng-on-*
+            nName = directiveNormalize(name);
+            if ((ngPrefixMatch = nName.match(NG_PREFIX_BINDING))) {
+              isNgAttr = ngPrefixMatch[1] === 'Attr';
+              isNgProp = ngPrefixMatch[1] === 'Prop';
+              isNgEvent = ngPrefixMatch[1] === 'On';
+
               name = name.replace(PREFIX_REGEXP, '')
-                .substr(8).replace(/_(.)/g, function(match, letter) {
+                .substr(4 + ngPrefixMatch[1].length).replace(/_(.)/g, function(match, letter) {
                   return letter.toUpperCase();
                 });
-            }
 
-            var multiElementMatch = ngAttrName.match(MULTI_ELEMENT_DIR_RE);
-            if (multiElementMatch && directiveIsMultiElement(multiElementMatch[1])) {
+            // support *-start / *-end multi element directives
+            } else if ((multiElementMatch = nName.match(MULTI_ELEMENT_DIR_RE)) && directiveIsMultiElement(multiElementMatch[1])) {
               attrStartName = name;
               attrEndName = name.substr(0, name.length - 5) + 'end';
               name = name.substr(0, name.length - 6);
             }
 
-            nName = directiveNormalize(name.toLowerCase());
-            attrsMap[nName] = name;
-            if (isNgAttr || !attrs.hasOwnProperty(nName)) {
+            if (isNgProp || isNgEvent) {
+              attrs[nName] = value;
+              attrsMap[nName] = attr.name;
```

Comment:
For reference, this is the difference between how `ngAttr` and `ngProp` attributes get mapped (and it looks reasonable to me):

```js
// HTML: x-ng-attr-my-form_action
attrsMap['myFormaction'] = 'my-formAction'

// HTML: x-ng-prop-my-form_action
attrsMap['ngPropMyForm_action'] = 'x-ng-prop-my-form_action'
```

Reply:
Should the `_`s not behave the same?

I'll try to add some tests with `_`s for both `ngAttr` and `ngProp`

Reply:
I've added a commit https://github.com/angular/angular.js/pull/16614/commits/c032255cbaa3ca9b748f4d4b4dc6fe70bb5286c8 (or two https://github.com/angular/angular.js/pull/16614/commits/7835fc36c65178ee6985f797e898c449aee35e81) to test this.

The result is a bit different then you assumed for the ngProp case:
```
// HTML: x-ng-attr-my-form_action
attrsMap['myFormaction'] = 'my-formAction'

// HTML: x-ng-prop-my-form_action
attrsMap['ngPropMyFormAction'] = 'x-ng-prop-my-form_action'
```

Notice the capital A for the ng-prop one... I think that is actually more correct then the ng-attr one. This is because we invoke `directiveNormalize` in two different ways:
* https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L2258 - on the raw attribute name
* https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L2274 - on the no-prefix attribute name, but in all lower case. This lower casing causes the mapping to be `myFormaction` instead of `myFormAction`

Is this a bug? Or a feature and we should be copying for `ng-prop`? :thinking:

Reply:
Yeah... the `element.attr('...')` and the real attribute name are correct, it's only the [nName](https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L2274) variable which is what is used on `$attrs` that is wrong.

Should we fix it though? If anyone is depending on these bad names it would be hard to find+fix if we fixed the bug. However it would be nice to fix the bug...

Reply:
Created https://github.com/angular/angular.js/issues/16624